### PR TITLE
Issue359

### PR DESCRIPTION
**Pull request by mreboud**
This fixes a few weaknesses of the expression parser:

- Expression observables can be defined at compile time
- Expression observables handle cached observables
- Exponentiation is now implemented

We also remove the, now useless, `make_observable_ratio` and `_sum` functions and convert the corresponding observables.

One point of #359 is still pending, namely the handling of `Unit` in the python interface of expression observables.

### CODE REVIEW

File: eos/utils/Makefile.am [hunk 1]
```diff
@@ -22,8 +22,13 @@ libeosutils_la_SOURCES = \
 	diagnostics.cc diagnostics.hh \
 	exception.cc exception.hh \
 	expression.cc expression.hh expression-fwd.hh \
+	expression-cacher.hh \
+	expression-cloner.hh \
+	expression-evaluator.hh \
+	expression-maker.hh \
 	expression-observable.cc expression-observable.hh \
-	expression-parser.hh expression-parser-impl.hh \
+	expression-parser.cc expressions-parser.hh expression-parser-impl.hh \
```
Comment (mreboud) [inline eos/utils/Makefile.am:30]: @dvandyk you changed my correct line and added a faulty `s` ^^

Comment (dvandyk) [reply eos/utils/Makefile.am]: Uff, I rebased onto what was already in gh/issue359. Maybe you did not commit and force push?

Comment (mreboud) [reply eos/utils/Makefile.am]: Ok, it is corrected :)

File: eos/b-decays/observables.cc [hunk 2]
```diff
@@ -285,23 +297,21 @@ namespace eos
                         std::make_tuple("q2_min", "q2_max"),
                         Options{ { "U", "c" }, {"q", "s"} }),
 
-                make_observable_ratio("B_s->D_slnu::R_D_s(q2)", R"(R_D_s(q^2))",
+                make_expression_observable("B_s->D_slnu::R_D_s(q2)", R"(R_D_s(q^2))",
```
Comment (mreboud) [inline eos/b-decays/observables.cc:300]: Double subscript there!

File: eos/b-decays/observables.cc [hunk 3]
```diff
@@ -285,23 +297,21 @@ namespace eos
...
-                        std::make_tuple("q2"),
-                        Options{ { "U", "c" }, {"q", "s"}, { "l", "mu" } }),
+                        R"(
+                        <<B_s->D_slnu::dBR/dq2;U=c,q=s,l=tau>>
+                        /
+                        <<B_s->D_slnu::dBR/dq2;U=c,q=s,l=mu>>
+                        )"),
 
-                make_observable_ratio("B_s->D_slnu::R_D_s", R"(R_D_s)",
+                make_expression_observable("B_s->D_slnu::R_D_s", R"(R_D_s)",
```
Comment (mreboud) [inline eos/b-decays/observables.cc:308]: ...and there too

**Review by dvandyk — APPROVED** [review]
This looks good. I'll rebase and merge with admin privileges.

**Pull request merged**